File: one_to_five_h/46-permutations.py
"""
    Leetcode 46
    From https://leetcode-cn.com/problems/permutations/
"""


def permute(nums):
    if not nums:
        return []
    res = []
    length = len(nums)

    def dfs(depth, path, visited):
        if depth == length:
            res.append(path)
            return

        for i in range(0, length):
            if visited[i]:
                continue
            visited[i] = True
            dfs(depth + 1, path + [nums[i]], visited)
            visited[i] = False
    visited = [False for i in range(length)]
    dfs(0, [], visited)
    return res

File: backtrack/46-permutations.go
package main

func permute(nums []int) [][]int {
	res := [][]int{}
	var dfs func([]int, int, int)
	dfs = func(path []int, visited, dep int) {
		if dep == len(nums) {
			tmp := make([]int, len(nums))
			copy(tmp, path)
			res = append(res, tmp)
			return
		}
		for idx, e := range nums {
			if visited>>idx&1 == 1 {
				continue
			}
			path = append(path, e)
			visited ^= 1 << idx
			dfs(path, visited, dep+1)
			path = path[:len(path)-1]
			visited ^= 1 << idx
		}
	}
	dfs([]int{}, 0, 0)
	return res
}
